src/kv/LockResolver.cc

```diff
@@ -305,6 +303,60 @@ void LockResolver::resolvePessimisticLock(Backoffer & bo, LockPtr lock, std::uno
     }
 }
 
+
+static std::pair<std::unordered_map<RegionVerID, Backoffer *>, std::vector<Backoffer>> genInnerBackoffer(
+    Backoffer & bo,
+    const std::unordered_map<RegionVerID, std::vector<std::string>> & keys_by_region)
+{
+    std::vector<Backoffer> data_holder;
+    std::unordered_map<RegionVerID, Backoffer *> res;
+    std::unordered_set<uint64_t> used_region_id;
+
+    data_holder.reserve(keys_by_region.size());
+
+    if (!bo.region_backoff_map)
+    {
+        for (const auto & pair : keys_by_region)
+        {
+            if (res.empty())
+            {
+                res.emplace(pair.first, &bo);
+            }
+            else
+            {
+                data_holder.emplace_back(kv::copNextMaxBackoff);
+                res.emplace(pair.first, &data_holder.back());
+            }
+        }
+        return {std::move(res), std::move(data_holder)};
+    }
+
+
+    for (const auto & pair : keys_by_region)
+    {
+        auto [it, ok] = bo.region_backoff_map->try_emplace(pair.first.id, kv::copNextMaxBackoff, bo.region_backoff_map);
+        if (ok)
+        {
+            res[pair.first] = &it->second;
+        }
+        else
+        {
+            if (used_region_id.count(pair.first.id))
```

Comment:
I think `keys_by_region` should have no duplicated `region_id`s, how about just throw exception here?

Reply:
never mind, I got the difference

Reply:
In TiDB's implementation, it just use bo.Fork to copy current bo, how about we follow the TiDB's behavior?
https://github.com/tikv/client-go/blob/master/txnkv/txnlock/lock_resolver.go#L1034

Reply:
The behavior of TiDB is unreasonable but it is simple for using.